# **DEMO 2**

Dataset: [Pokemon with stats (link)](https://www.kaggle.com/datasets/abcsds/pokemon)

Selection Rationale: I choose this dataset because I was recently playing the new nintendo switch game

___

### Analysis Objectives & Plan
1. You want to become a strong pokemon gym leader using the stats file! 
- First reset the index to the 'Name' of pokemon within the file.
- Condense the number of columns to visualize the pokemon type and total basepoints ('Total')
- Remove any unnecessary columns in the dataset.
2. Most gym leaders have a team make up of a particular pokemon type. Sort your list by pokemon type. 
- Sort the pokemon with the highest HP (health points) and type.
- Sort pokemon with highest overall (Total) points & create a list based on strongest pokemon in a descedning order. 
- Now choose a type based on the top few strongest. Organize the list to have your chosen type and highest Total points.
3. The strongest pokemon within certain types may not be possible to capture or aviable in all games. Generate the list of realistic pokemon you can use. 
- Remove options you likely won't be able to obtain within the game (i.e. legendary)
- Provide descriptive statistics for your new list. 
- What is the mean values of the new list
- Based on highest speed stats choose your team of 6 pokemon

___

### Import Libraries 

In [127]:
# import necessary libraries 
import pandas as pd 

___

## Import and Explore Data
- import dataset and assign to a variable named df
- print/output the following (as pythonically as possible):
    - DataFrame's shape
    - Column names 
    - First 5 rows 

In [128]:
# import data from csv file
df = pd.read_csv('Pokemon2.csv')

In [129]:
df 

#                       Name    Type 1    Type 2  Total   HP  Attack  \
0      1                  Bulbasaur     Grass    Poison    318   45      49   
1      2                    Ivysaur     Grass    Poison    405   60      62   
2      3                   Venusaur     Grass    Poison    525   80      82   
3      3      VenusaurMega Venusaur     Grass    Poison    625   80     100   
4      4                 Charmander      Fire       NaN    309   39      52   
5      5                 Charmeleon      Fire       NaN    405   58      64   
6      6                  Charizard      Fire    Flying    534   78      84   
7      6  CharizardMega Charizard X      Fire    Dragon    634   78     130   
8      6  CharizardMega Charizard Y      Fire    Flying    634   78     104   
9      7                   Squirtle     Water       NaN    314   44      48   
10     8                  Wartortle     Water       NaN    405   59      63   
11     9                  Blastoise     Water       NaN    530   79      83   
12     9    BlastoiseMega Blastoise     Water       NaN    630   79     103   
13    10                   Caterpie       Bug       NaN    195   45      30   
14    11                    Metapod       Bug       NaN    205   50      20   
15    12                 Butterfree       Bug    Flying    395   60      45   
16    13                     Weedle       Bug    Poison    195   40      35   
17    14                     Kakuna       Bug    Poison    205   45      25   
18    15                   Beedrill       Bug    Poison    395   65      90   
19    15      BeedrillMega Beedrill       Bug    Poison    495   65     150   
20    16                     Pidgey    Normal    Flying    251   40      45   
21    17                  Pidgeotto    Normal    Flying    349   63      60   
22    18                    Pidgeot    Normal    Flying    479   83      80   
23    18        PidgeotMega Pidgeot    Normal    Flying    579   83      80   
24    19                    Rattata    Normal       NaN    253   30      56   
25    20                   Raticate    Normal       NaN    413   55      81   
26    21                    Spearow    Normal    Flying    262   40      60   
27    22                     Fearow    Normal    Flying    442   65      90   
28    23                      Ekans    Poison       NaN    288   35      60   
29    24                      Arbok    Poison       NaN    438   60      85   
30    25                    Pikachu  Electric       NaN    320   35      55   
31    26                     Raichu  Electric       NaN    485   60      90   
32    27                  Sandshrew    Ground       NaN    300   50      75   
33    28                  Sandslash    Ground       NaN    450   75     100   
34    29                   Nidoran♀    Poison       NaN    275   55      47   
35    30                   Nidorina    Poison       NaN    365   70      62   
36    31                  Nidoqueen    Poison    Ground    505   90      92   
37    32                   Nidoran♂    Poison       NaN    273   46      57   
38    33                   Nidorino    Poison       NaN    365   61      72   
39    34                   Nidoking    Poison    Ground    505   81     102   
40    35                   Clefairy     Fairy       NaN    323   70      45   
41    36                   Clefable     Fairy       NaN    483   95      70   
42    37                     Vulpix      Fire       NaN    299   38      41   
43    38                  Ninetales      Fire       NaN    505   73      76   
44    39                 Jigglypuff    Normal     Fairy    270  115      45   
45    40                 Wigglytuff    Normal     Fairy    435  140      70   
46    41                      Zubat    Poison    Flying    245   40      45   
47    42                     Golbat    Poison    Flying    455   75      80   
48    43                     Oddish     Grass    Poison    320   45      50   
49    44                      Gloom     Grass    Poison  

In [130]:
# df shape
print(df.shape)

(800, 13)


In [131]:
# df columns 
print(df.columns)

Index(['#', 'Name', 'Type 1', 'Type 2', 'Total', 'HP', 'Attack', 'Defense',
       'Sp. Atk', 'Sp. Def', 'Speed', 'Generation', 'Legendary'],
      dtype='object')


In [132]:
# output first 5 rows 
print(df.head())

   #                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False  


___

### Import and Explain Tasks 

___

##### Objective 1

In [133]:
# Step 1. We don't really care about the pokemon number to reassign Name to be the first column
df = df.set_index('Name')


In [134]:
df

#    Type 1    Type 2  Total   HP  Attack  \
Name                                                                     
Bulbasaur                    1     Grass    Poison    318   45      49   
Ivysaur                      2     Grass    Poison    405   60      62   
Venusaur                     3     Grass    Poison    525   80      82   
VenusaurMega Venusaur        3     Grass    Poison    625   80     100   
Charmander                   4      Fire       NaN    309   39      52   
Charmeleon                   5      Fire       NaN    405   58      64   
Charizard                    6      Fire    Flying    534   78      84   
CharizardMega Charizard X    6      Fire    Dragon    634   78     130   
CharizardMega Charizard Y    6      Fire    Flying    634   78     104   
Squirtle                     7     Water       NaN    314   44      48   
Wartortle                    8     Water       NaN    405   59      63   
Blastoise                    9     Water       NaN    530   79      83   
BlastoiseMega Blastoise      9     Water       NaN    630   79     103   
Caterpie                    10       Bug       NaN    195   45      30   
Metapod                     11       Bug       NaN    205   50      20   
Butterfree                  12       Bug    Flying    395   60      45   
Weedle                      13       Bug    Poison    195   40      35   
Kakuna                      14       Bug    Poison    205   45      25   
Beedrill                    15       Bug    Poison    395   65      90   
BeedrillMega Beedrill       15       Bug    Poison    495   65     150   
Pidgey                      16    Normal    Flying    251   40      45   
Pidgeotto                   17    Normal    Flying    349   63      60   
Pidgeot                     18    Normal    Flying    479   83      80   
PidgeotMega Pidgeot         18    Normal    Flying    579   83      80   
Rattata                     19    Normal       NaN    253   30      56   
Raticate                    20    Normal       NaN    413   55      81   
Spearow                     21    Normal    Flying    262   40      60   
Fearow                      22    Normal    Flying    442   65      90   
Ekans                       23    Poison       NaN    288   35      60   
Arbok                       24    Poison       NaN    438   60      85   
Pikachu                     25  Electric       NaN    320   35      55   
Raichu                      26  Electric       NaN    485   60      90   
Sandshrew                   27    Ground       NaN    300   50      75   
Sandslash                   28    Ground       NaN    450   75     100   
Nidoran♀                    29    Poison       NaN    275   55      47   
Nidorina                    30    Poison       NaN    365   70      62   
Nidoqueen                   31    Poison    Ground    505   90      92   
Nidoran♂                    32    Poison       NaN    273   46      57   
Nidorino                    33    Poison       NaN    365   61      72   
Nidoking                    34    Poison    Ground    505   81     102   
Clefairy                    35     Fairy       NaN    323   70      45   
Clefable                    36     Fairy       NaN    483   95      70   
Vulpix                      37      Fire       NaN    299   38      41   
Ninetales                   38      Fire       NaN    505   73      76   
Jigglypuff                  39    Normal     Fairy    270  115      45   
Wigglytuff                  40    Normal     Fairy    435  140      70   
Zubat                       41    Poison    Flying    245   40      45   
Golbat                      42    Poison    Flying    455   75      80   
Oddish                      43     Grass    Poison    320   45      50   
Gloom                       44     Grass    Poison    395   60      65   
Vileplume                   45     Grass    Poison    490   75      80   
Paras                       46       Bug     Grass    285   35      70   
Parasect                    47   

In [135]:
#Step 2. We wish to see primarily the pokemon type and total base points. Index the table to show these in a simplier view
df.iloc[:-1, 0:4] 

#    Type 1    Type 2  Total
Name                                                     
Bulbasaur                    1     Grass    Poison    318
Ivysaur                      2     Grass    Poison    405
Venusaur                     3     Grass    Poison    525
VenusaurMega Venusaur        3     Grass    Poison    625
Charmander                   4      Fire       NaN    309
Charmeleon                   5      Fire       NaN    405
Charizard                    6      Fire    Flying    534
CharizardMega Charizard X    6      Fire    Dragon    634
CharizardMega Charizard Y    6      Fire    Flying    634
Squirtle                     7     Water       NaN    314
Wartortle                    8     Water       NaN    405
Blastoise                    9     Water       NaN    530
BlastoiseMega Blastoise      9     Water       NaN    630
Caterpie                    10       Bug       NaN    195
Metapod                     11       Bug       NaN    205
Butterfree                  12       Bug    Flying    395
Weedle                      13       Bug    Poison    195
Kakuna                      14       Bug    Poison    205
Beedrill                    15       Bug    Poison    395
BeedrillMega Beedrill       15       Bug    Poison    495
Pidgey                      16    Normal    Flying    251
Pidgeotto                   17    Normal    Flying    349
Pidgeot                     18    Normal    Flying    479
PidgeotMega Pidgeot         18    Normal    Flying    579
Rattata                     19    Normal       NaN    253
Raticate                    20    Normal       NaN    413
Spearow                     21    Normal    Flying    262
Fearow                      22    Normal    Flying    442
Ekans                       23    Poison       NaN    288
Arbok                       24    Poison       NaN    438
Pikachu                     25  Electric       NaN    320
Raichu                      26  Electric       NaN    485
Sandshrew                   27    Ground       NaN    300
Sandslash                   28    Ground       NaN    450
Nidoran♀                    29    Poison       NaN    275
Nidorina                    30    Poison       NaN    365
Nidoqueen                   31    Poison    Ground    505
Nidoran♂                    32    Poison       NaN    273
Nidorino                    33    Poison       NaN    365
Nidoking                    34    Poison    Ground    505
Clefairy                    35     Fairy       NaN    323
Clefable                    36     Fairy       NaN    483
Vulpix                      37      Fire       NaN    299
Ninetales                   38      Fire       NaN    505
Jigglypuff                  39    Normal     Fairy    270
Wigglytuff                  40    Normal     Fairy    435
Zubat                       41    Poison    Flying    245
Golbat                      42    Poison    Flying    455
Oddish                      43     Grass    Poison    320
Gloom                       44     Grass    Poison    395
Vileplume                   45     Grass    Poison    490
Paras                       46       Bug     Grass    285
Parasect                    47       Bug     Grass    405
Venonat                     48       Bug    Poison    305
Venomoth                    49       Bug    Poison    450
Diglett                     50    Ground       NaN    265
Dugtrio                     51    Ground       NaN    405
Meowth                      52    Normal       NaN    290
Persian                     53    Normal       NaN    440
Psyduck                     54     Water       NaN    320
Golduck                     55     Water       NaN    500
Mankey                      56  Fighting       NaN    305
Primeape                    57  Fighting       NaN    455
Growlithe                   58      Fire       NaN    350
Arcanine                    59      Fire       NaN    555
Poliwag                     60     Water       NaN    300
Poliwhirl                   61     Water       NaN    385
Poliwrath                  

In [136]:
#Step 3. We don't care which generation the pokemon is from, so we can drop this column
df = df.drop(columns = 'Generation')

___

##### Objective 2

In [137]:
# step 1. We want to see the pokemon with the highest HP sorted by Type. Sort the table to show this
df_sorted = df.sort_values(['Type 1', 'HP'], ascending=[1,0])

In [138]:
print(df_sorted)

                             #    Type 1    Type 2  Total   HP  Attack  \
Name                                                                     
Yanmega                    469       Bug    Flying    515   86      76   
Volcarona                  637       Bug      Fire    550   85      60   
Heracross                  214       Bug  Fighting    500   80     125   
HeracrossMega Heracross    214       Bug  Fighting    600   80     185   
Accelgor                   617       Bug       NaN    495   80      70   
Vivillon                   666       Bug    Flying    411   80      52   
Kricketune                 402       Bug       NaN    384   77      85   
Forretress                 205       Bug     Steel    465   75      90   
Leavanny                   542       Bug     Grass    500   75     103   
Genesect                   649       Bug     Steel    600   71     120   
Venomoth                    49       Bug    Poison    450   70      65   
Scyther                    123       B

In [139]:
#Step 2. We want pokemon with high overall (Total) points, create a list based on strongest pokemon in a descedning order. 
condition = df['Total'] > 500
sorted_df = df[condition].sort_values(by='Total', ascending=False)
print(sorted_df)

                             #    Type 1    Type 2  Total   HP  Attack  \
Name                                                                     
RayquazaMega Rayquaza      384    Dragon    Flying    780  105     180   
MewtwoMega Mewtwo X        150   Psychic  Fighting    780  106     190   
MewtwoMega Mewtwo Y        150   Psychic       NaN    780  106     150   
GroudonPrimal Groudon      383    Ground      Fire    770  100     180   
KyogrePrimal Kyogre        382     Water       NaN    770  100     150   
Arceus                     493    Normal       NaN    720  120     120   
SalamenceMega Salamence    373    Dragon    Flying    700   95     145   
KyuremWhite Kyurem         646    Dragon       Ice    700  125     120   
KyuremBlack Kyurem         646    Dragon       Ice    700  125     170   
LatiosMega Latios          381    Dragon   Psychic    700   80     130   
GarchompMega Garchomp      445    Dragon    Ground    700  108     170   
LatiasMega Latias          380    Drag

___

In [140]:
#Step 3. You decided to choose psychic pokemon as your type to train. Organize the list to have psychic type and highest Total points.
psychic_condition = df['Type 1'] == 'Psychic'
hp_condition = df['Total'] > 500
combined_condition = psychic_condition & hp_condition
psychic_pokemon_high_hp = df[combined_condition].sort_values(by='Total', ascending=False)
print(psychic_pokemon_high_hp)

                           #   Type 1    Type 2  Total   HP  Attack  Defense  \
Name                                                                           
MewtwoMega Mewtwo X      150  Psychic  Fighting    780  106     190      100   
MewtwoMega Mewtwo Y      150  Psychic       NaN    780  106     150       70   
HoopaHoopa Unbound       720  Psychic      Dark    680   80     160       60   
Lugia                    249  Psychic    Flying    680  106      90      130   
Mewtwo                   150  Psychic       NaN    680  106     110       90   
GalladeMega Gallade      475  Psychic  Fighting    618   68     165       95   
GardevoirMega Gardevoir  282  Psychic     Fairy    618   68      85       65   
DeoxysDefense Forme      386  Psychic       NaN    600   50      70      160   
HoopaHoopa Confined      720  Psychic     Ghost    600   80     110       60   
Victini                  494  Psychic      Fire    600  100     100      100   
Cresselia                488  Psychic   

##### Objective 3

In [141]:
# step 1. Realistically, you will not have legendary pokemon on your team. Filter your new list to remove legendary pokemon.
new_table_rows = df[(df['Type 1'] == 'Psychic') & (df['Total'] > 500) & (df['Legendary'] == False)]
print(new_table_rows)

                           #   Type 1    Type 2  Total   HP  Attack  Defense  \
Name                                                                           
AlakazamMega Alakazam     65  Psychic       NaN    590   55      50       65   
Mew                      151  Psychic       NaN    600  100     100      100   
Espeon                   196  Psychic       NaN    525   65      65       60   
Celebi                   251  Psychic     Grass    600  100     100      100   
Gardevoir                282  Psychic     Fairy    518   68      65       65   
GardevoirMega Gardevoir  282  Psychic     Fairy    618   68      85       65   
Gallade                  475  Psychic  Fighting    518   68     125       65   
GalladeMega Gallade      475  Psychic  Fighting    618   68     165       95   
Cresselia                488  Psychic       NaN    600  120      70      120   

                         Sp. Atk  Sp. Def  Speed  Legendary  
Name                                                     

In [142]:
# step 2. What are the descriptive statistics of your new list?
df = new_table_rows
df.describe()

#       Total          HP      Attack     Defense     Sp. Atk  \
count    9.000000    9.000000    9.000000    9.000000    9.000000    9.000000   
mean   296.111111  576.333333   79.111111   91.666667   81.666667  111.111111   
std    153.194684   42.982555   21.848595   35.881750   22.079402   40.833333   
min     65.000000  518.000000   55.000000   50.000000   60.000000   65.000000   
25%    196.000000  525.000000   68.000000   65.000000   65.000000   75.000000   
50%    282.000000  600.000000   68.000000   85.000000   65.000000  100.000000   
75%    475.000000  600.000000  100.000000  100.000000  100.000000  130.000000   
max    488.000000  618.000000  120.000000  165.000000  120.000000  175.000000   

          Sp. Def       Speed  
count    9.000000    9.000000  
mean   111.111111  101.666667  
std     14.743172   21.505813  
min     95.000000   80.000000  
25%    100.000000   85.000000  
50%    115.000000  100.000000  
75%    115.000000  110.000000  
max    135.000000  150.000000

In [143]:
# step 3. What is the mean of your new list?
df.mean()

/var/folders/6f/m73m670n5b7f2f4_2zpzz6hh0000gn/T/ipykernel_3565/2172828939.py:2: FutureWarning: The default value of numeric_only in DataFrame.mean is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  df.mean()


#            296.111111
Total        576.333333
HP            79.111111
Attack        91.666667
Defense       81.666667
Sp. Atk      111.111111
Sp. Def      111.111111
Speed        101.666667
Legendary      0.000000
dtype: float64

In [144]:
#Step 4. Based on highest speed stats choose your team of 6 pokemon.
my_team = ['AlakazamMega Alakazam', 'Espeon', 'Celebi', 'Mew', 'GardevoirMega Gardevoir', 'GalladeMega Gallade']

In [145]:
df.loc[my_team]

#   Type 1    Type 2  Total   HP  Attack  Defense  \
Name                                                                           
AlakazamMega Alakazam     65  Psychic       NaN    590   55      50       65   
Espeon                   196  Psychic       NaN    525   65      65       60   
Celebi                   251  Psychic     Grass    600  100     100      100   
Mew                      151  Psychic       NaN    600  100     100      100   
GardevoirMega Gardevoir  282  Psychic     Fairy    618   68      85       65   
GalladeMega Gallade      475  Psychic  Fighting    618   68     165       95   

                         Sp. Atk  Sp. Def  Speed  Legendary  
Name                                                         
AlakazamMega Alakazam        175       95    150      False  
Espeon                       130       95    110      False  
Celebi                       100      100    100      False  
Mew                          100      100    100      False  
GardevoirMega Gardevoir      165      135    100      False  
GalladeMega Gallade           65      115    110      False

___

##### Conclusions

Based on filtering for total stat points, removing uncessasy columns within the dataset, sorting by type, and sorting by ascending statistics we were able to generate a list based on strong types & unltimating deciding on a psychic type list. Additional conditions based on legendary status allowed us to generate a team of 6 would be realisitically capture & train within a game to run a psychic style gym. 

We narrowed down our list in the end & could make our final choice based on another stat (speed) since the list was condensed. We could then generate basic descriptive stats of this list with the MEAN showing the most useful information. This indicated our teams mean total base stats was 576.333333, which is good setting our threshold at >500 and the highest in the game being 780 (for a legendary). 